contracts/Joint.sol

```diff
@@ -159,85 +185,302 @@ contract Joint {
-            swapReward(rewardProfit);
+        if (returnFunds) {
+            _returnLooseToProviders();
         }
+    }
 
+    function adjustPosition(bool reinvest) external onlyProviders {
         // No capital, nothing to do
-        if (balanceOfA() == 0 && balanceOfB() == 0) {
+        if (balanceOfA() == 0 || balanceOfB() == 0) {
             return;
         }
 
         if (reinvest) {
-            createLP();
+            require(
+                balanceOfStake() == 0 &&
+                    balanceOfPair() == 0 &&
+                    investedA == 0 &&
+                    investedB == 0
+            ); // don't create LP if we are already invested
+
+            (investedA, investedB, ) = createLP();
             depositLP();
+        }
+    }
+
+    function estimatedTotalAssetsAfterBalance()
+        public
+        view
+        returns (uint256 _aBalance, uint256 _bBalance)
+    {
+        uint256 rewardsPending = pendingReward();
+
+        (_aBalance, _bBalance) = balanceOfTokensInLP();
+
+        if (reward == tokenA) {
+            _aBalance = _aBalance.add(rewardsPending);
+        } else if (reward == tokenB) {
+            _bBalance = _bBalance.add(rewardsPending);
+        } else if (rewardsPending != 0) {
+            address swapTo = findSwapTo(reward);
+            uint256[] memory outAmounts =
+                IUniswapV2Router02(router).getAmountsOut(
+                    rewardsPending,
+                    getTokenOutPath(reward, swapTo)
+                );
+            if (swapTo == tokenA) {
+                _aBalance = _aBalance.add(outAmounts[outAmounts.length - 1]);
+            } else if (swapTo == tokenB) {
+                _bBalance = _bBalance.add(outAmounts[outAmounts.length - 1]);
+            }
+        }
+
+        (address sellToken, uint256 sellAmount) =
+            calculateSellToBalance(_aBalance, _bBalance, investedA, investedB);
+
+        (uint256 reserveA, uint256 reserveB) = getReserves();
+
+        if (sellToken == tokenA) {
+            uint256 buyAmount =
+                IUniswapV2Router02(router).getAmountOut(
+                    sellAmount,
+                    reserveA,
+                    reserveB
+                );
+            _aBalance = _aBalance.sub(sellAmount);
+            _bBalance = _bBalance.add(buyAmount);
+        } else if (sellToken == tokenB) {
+            uint256 buyAmount =
+                IUniswapV2Router02(router).getAmountOut(
+                    sellAmount,
+                    reserveB,
+                    reserveA
+                );
+            _bBalance = _bBalance.sub(sellAmount);
+            _aBalance = _aBalance.add(buyAmount);
+        }
+
+        _aBalance = _aBalance.add(balanceOfA());
+        _bBalance = _bBalance.add(balanceOfB());
+    }
+
+    function estimatedTotalAssetsInToken(address token)
+        external
+        view
+        returns (uint256 _balance)
+    {
+        require(token == tokenA || token == tokenB);
```

Comment:
do we really need this require?

maybe replace else with else if (token == tokenB) and return 0 otherwise